### ISSUE TITLE
Multiple SSH Tunnels from Multiple Systems

### ISSUE THREAD

**Issue by ommikadam** [Genymobile/scrcpy#2292]
How I can connect to adb server where multiple devices are connected with multiple systems at the same time.
For example, I have systems A, B & C where A is the server and B & C are remote machines. So If I added smartphones on System A, and try to connect from System B it works perfectly fine with SSH Tunnel, but at the same time if I try to connect from System C the ADB server and smartphones are disconnected on the server also the SSH Tunnel says: channel 4: open failed: connect failed: connection refused.

adb versions on all the 3 machines are same - 1.0.39
Scrcpy versions are same - 1.12.1

I've tried with local port forwarding but scrcpy gives the error of --force-adb-forward is unrecognized.

Please help

**Comment by rom1v** [comment]
What exact commands did you execute to enable the SSH tunnels and run scrcpy?

(also note: https://stackoverflow.com/a/42375902/1987178)

**Comment by ommikadam** [comment]
Please find the procedure from start to end which I followed

System A ( Server)
1. adb start-server
2. adb -d tcpip 5555
3. disconnected the phone from usb
3. adb connect xxx.xxx.xxx.xxx:5555
4. adb devices (Shows the device)
5. adb -d tcpip 5556
6. disconnected the phone from usb
7. adb connect xxx.xxx.xxx.xxx:5556
8. adb devices (Shows the both devices)
9. Now open the another window 
10. scrcpy --serial xxxx.xxxx.xxxx.xxxx:5555 (Device 01 comes up on server)
11. scrcpy --serial xxxx.xxxx.xxxx.xxxx:5556 (Device 02 comes up on server)

System B (Remote side)
1. adb kill-server
2. ssh -CN -L5037:localhost:5037 -R27183:localhost:27183 administrator@Server_IP
3. Open another window
4. scrcpy --serial xxxx.xxxx.xxxx.xxxx:5555 (Device 01 comes up on remote System B)

System C (Remote Side)
1. adb kill-server
2. ssh -CN -L5037:localhost:5037 -R27183:localhost:27183 administrator@Server_IP
3. Open another window
4. scrcpy --serial xxxx.xxxx.xxxx.xxxx:5556 (Device 02 fails to come up on remote System B)
5. channel 4: open failed: connect failed: connection refused. - Error in ssh tunnel window

Also, the adb server disconnects all the connected devices,s and its down.

Tried the same with ssh -CN -L5037:localhost:5037 -L27183:localhost:27183 administrator@Server_IP but after that scrcpy --force-adb-forward this command is an unrecognized error.

please help

**Comment by rom1v** [comment]
On system A, you must use another port or system C.

In practice, on B, execute:
```bash
ssh -CN -L5037:localhost:5037 -R27183:localhost:27183 administrator@Server_IP
scrcpy -P 27183
```

and on C:
```bash
ssh -CN -L5037:localhost:5037 -R27184:localhost:27184 administrator@Server_IP
scrcpy -P 27184
```

> scrcpy --force-adb-forward this command is an unrecognized error.

Because you use an old version (1.12.1).

**Comment by ommikadam** [comment]
Hi,

I am
![Capture](https://user-images.githubusercontent.com/80241548/116907125-c9f34a00-ac5e-11eb-8235-f37cba4bafbf.PNG)
 appreciating your prompt responses and now I've upgraded the scrcpy to the latest version i.e, 1.17, and when I try what you suggested it gives me error attached is the screenshot for your reference.

**Comment by rom1v** [comment]
> it gives me error attached is the screenshot

(please copy-paste text instead, it is indexed and can be searched easily)

> Segmentation fault

Maybe: https://github.com/Genymobile/scrcpy/issues/2053#issuecomment-801895765

> sudo scrcpy …

It is an error to run scrcpy as root. If you haveto run it as root, you probably build it with wrong permissions. You should uninstall and rebuild correctly.

**Comment by ommikadam** [comment]
Thanks for the update I've removed the package and installed it correctly but now it works on system A but it fails on system B.
I've created different ports for the same and gives below error. 
For the troubleshooting i checked the open ports on server there were no port 27194 but after command its there and says already the port is bind

scrcpy --port 27194 --serial 192.168.4.137:5556
INFO: scrcpy 1.17 <https://github.com/Genymobile/scrcpy>
/usr/local/share/scrcpy/scrcpy-server:...shed. 1.8 MB/s (34930 bytes in 0.019s)
error: more than one device/emulator
ERROR: "adb reverse" returned with value 1
WARN: 'adb reverse' failed, fallback to 'adb forward'
error: cannot bind listener: Address already in use
ERROR: "adb forward" returned with value 1
ERROR: Could not forward port 27194
free(): double free detected in tcache 2
Aborted (core dumped)

**Comment by rom1v** [comment]
> error: cannot bind listener: Address already in use
> WARN: 'adb reverse' failed, fallback to 'adb forward'

OK, so over wifi, you can't `adb reverse` on your device (see #5). This is solved on earlier Android versions.
But since you enabled a SSH tunnel with remote redirection, your server A could not bind your port (it's already taken by ssh).

So instead, on B (notice `-L` instead of `-R`):
```bash
ssh -CN -L5037:localhost:5037 -L27183:localhost:27183 administrator@Server_IP
scrcpy -p 27183 -s xx.xx.xx.xx:5556 --force-adb-forward
```
And on C:
```
ssh -CN -L5037:localhost:5037 -L27184:localhost:27184 administrator@Server_IP
scrcpy -p 27184 -s yy.yy.yy.yy:5556 --force-adb-forward
```

_Note that you could have use 5555 for both devices, there is no need to use another port, they have a different IP address._

> ```
> free(): double free detected in tcache 2
> Aborted (core dumped)
> ```

Oh, thank you, there was a double-free on error there when a serial is provided. Fixed on `dev` by 1b9dcce23c8f52aa1ba97fe30fad83930c60578d. (but you don't care for your problem, it's just a consequence of your issue)

**Comment by ommikadam** [comment]
Thanks, issue has been resolved....
